internal/parser/parser.go

```diff
@@ -206,8 +209,8 @@ func findPort(svc *corev1.Service, wantPort kongstate.PortDef) (*corev1.ServiceP
 }
 
 func getUpstreams(
-	log logrus.FieldLogger, s store.Storer, serviceMap map[string]kongstate.Service) []kongstate.Upstream {
-	var upstreams []kongstate.Upstream
+	log logrus.FieldLogger, s store.Storer, serviceMap map[string]kongstate.Service) map[string]kongstate.Upstream {
+	upstreams := make(map[string]kongstate.Upstream)
```

Comment:
regarding the use of a map here for de-duplication:

Some use cases may end up having very large numbers of `kongstate.Upstreams`. This code path is potentially one of our critical paths, it's being _constantly_ addressed under normal usage of the software.

Using a map for de-duplication is not generally a bad idea but I think we will have unnecessary memory proliferation with this implementation because we're going to be putting a lot of extra copies of each `kongstate.Upstream` contents on the heap (we store one copy in the map, and then the caller is copying those values into a separate slice).

For performance reasons I think we should refactor this implementation in the following ways:

1. initialize all slices and maps with a predefined underlying array size, as this will reduce the amount of work the Go runtime has do allocating new heap and copying array contents under the hood at runtime.
  a. for maps: `a := make(map[string]struct{}, len(serviceMap))`
  b. for slices: `a := make([]kongstate.Upstream, 0, len(serviceMap))`
2. use one of several potential options for a more memory and CPU efficient filtration and delivery of the list of upstreams:
  a. use a `map[string]struct{}` which will de-duplicate based on the unique names, and then return a simple slice
  b. use a `map[kongstate.Upstream]struct{}`
  c. implement a `Set` type for `kongstate` objects (using `2.b` under the hood) similar to `k8s.io/apimachinery/pkg/util/sets` (or use some existing 3rd party `Set` library type)

Each of the above has its advantages and disadvantages.

- `2.a` is a _good_ option due to the fact that we structurally need to deliver the `kongstate.Upstreams` as a slice to the existing calling code, but uses slightly more memory and CPU

- If we're willing to spend the time to do a larger refactor of the code where the calling code would happily operate on a `map[kongstate.Upstream]struct{}` `2.b` is the most performant and efficient of the three

- `2.c` is just a formalization of `2.b` which adds the value of making a general purpose implementation and improvements to overall code maintainability on top of the runtime performance improvements

Reply:
Will f5f2ab5f (only store the name string in the map, append kong.Upstream to the return slice if the name is not already present) work?

I initially tried 2.b in 
[attempt.diff.txt](https://github.com/Kong/kubernetes-ingress-controller/files/6994586/attempt.diff.txt) but it complains about it being an invalid map key type, so I don't think that's an option barring a unpleasant refactor of that type.

Reply:
Yes, the `2.a` option is a good starting point. I think it's fair to start there for this iteration and we can assess whether greater performance in this area is needed over time.